File: 벤치마크/go/sort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func partition(arr []int, lo, hi int) int {
	pivot := arr[hi]
	i := lo - 1
	for j := lo; j < hi; j++ {
		if arr[j] <= pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[hi] = arr[hi], arr[i+1]
	return i + 1
}

func quicksort(arr []int, lo, hi int) {
	if lo < hi {
		p := partition(arr, lo, hi)
		quicksort(arr, lo, p-1)
		quicksort(arr, p+1, hi)
	}
}

func main() {
	n := 1000000
	arr := make([]int, n)
	rng := rand.New(rand.NewSource(42))
	for i := 0; i < n; i++ {
		arr[i] = int(rng.Int31())
	}

	start := time.Now()
	quicksort(arr, 0, n-1)
	elapsed := time.Since(start).Seconds() * 1000

	fmt.Printf("sorted[0]=%d sorted[999999]=%d\n", arr[0], arr[999999])
	fmt.Printf("RESULT:%.3f\n", elapsed)
}

File: 벤치마크/rust/sort.rs
use std::time::Instant;

fn partition(arr: &mut [i32], lo: usize, hi: usize) -> usize {
    let pivot = arr[hi];
    let mut i = lo as isize - 1;
    for j in lo..hi {
        if arr[j] <= pivot {
            i += 1;
            arr.swap(i as usize, j);
        }
    }
    arr.swap((i + 1) as usize, hi);
    (i + 1) as usize
}

fn quicksort(arr: &mut [i32], lo: isize, hi: isize) {
    if lo < hi {
        let p = partition(arr, lo as usize, hi as usize);
        quicksort(arr, lo, p as isize - 1);
        quicksort(arr, p as isize + 1, hi);
    }
}

fn main() {
    let n = 1_000_000usize;
    let mut arr = Vec::with_capacity(n);

    // Simple LCG matching C's srand(42)/rand() pattern
    let mut seed: u32 = 42;
    for _ in 0..n {
        seed = seed.wrapping_mul(1103515245).wrapping_add(12345);
        arr.push(((seed >> 16) & 0x7fff) as i32);
    }

    let start = Instant::now();
    quicksort(&mut arr, 0, (n - 1) as isize);
    let ms = start.elapsed().as_secs_f64() * 1000.0;

    println!("sorted[0]={} sorted[999999]={}", arr[0], arr[999999]);
    println!("RESULT:{:.3}", ms);
}
